### DISCUSSION TITLE
Ubuntu For Windows Compatibility?

### DISCUSSION THREAD

**Discussion by fizzy123** [leolabs/ableton-js#90] [Q&A]
Hi! I'm trying to run a script that uses ableton-js in Ubuntu For Windows to communicate with the Ableton instance runing on the same Windows machine. 

The script itself is very simple - i'm just trying to verify connectivity:
```
const { Ableton } = require("ableton-js");
const ableton = new Ableton();

(async () => {
  let remover = await ableton.song.addListener("current_song_time", (async (time) => {
    console.log(1)
  }));
  await remover()
  ableton.song.addListener("current_song_time", (async (time) => {
    console.log(2)
  }));
  console.log("all set")
})()
```

This results in the following error:
```
nobel@DESKTOP-3MMEU6G:~/autodj$ node test.js
/home/nobel/autodj/node_modules/ableton-js/index.js:105
        var _this = _super.call(this, message) || this;
                           ^

Error: The command song.add_listener({"prop":"current_song_time","eventId":"44bf1854-71cd-4898-ae27-1f06a44cb5fb"}) timed out after 2000 ms. Please make sure that Ableton is running and that you have the latest version of AbletonJS' midi script installed and renamed to "AbletonJS", listening on port 39041 and sending on port 39031.
    at new TimeoutError (/home/nobel/autodj/node_modules/ableton-js/index.js:105:28)
    at Timeout._onTimeout (/home/nobel/autodj/node_modules/ableton-js/index.js:269:33)
    at listOnTimeout (node:internal/timers:573:17)
    at process.processTimers (node:internal/timers:514:7) {
  payload: {
    uuid: 'ad40d5ec-21a6-4c3e-b2cc-854e51687617',
    ns: 'song',
    nsid: undefined,
    name: 'add_listener',
    args: {
      prop: 'current_song_time',
      nsid: undefined,
      eventId: '44bf1854-71cd-4898-ae27-1f06a44cb5fb'
    }
  }
}
```

I'm guessing I've got to set up some kind of port forwarding between the Ubuntu and Windows systems? I was about to try port forwarding on 39041 and 39031 but then I read in the readme:
```
Both the client and the server bind to a random available port and store that port in a local file so the other side knows which port to send messages to.
```
Which makes it seem like the solution might be more complicated? Do I need to do some more configuration so that this local file can be accessed by both systems? If so, can anyone point me to the part of the code that I might need to update or modify? If code changes are required, I'd be happy to put together a PR to add this functionality as a configuration option!

**Comment by leolabs** [comment]
It seems that you're running an older version of the library which uses the fixed ports 39031 and 39041, but the newer version indeed uses random ports to avoid collisions with other programs on your machine. 

I don't have any experience with Ubuntu for Windows, unfortunately, so I can't help you with the port forwarding, but if you add a logger to the Ableton instance, it will tell you where the port files are located. As long as the ports are free, both ableton-js and its Live plugin will try to reuse the ports stored in these files. You can then try to forward those ports between Windows and the Ubuntu subsystem.

```ts
const ableton = new Ableton({ logger: console });
await ableton.start();
```

The output should look something like this:

```
Checking if a stored port exists {
  file: '/var/folders/b2/qgy9m6694wl9bqwmgpsnpdwm0000gn/T/ableton-js-client.port'
}
Trying to bind to the most recent port { port: 54935 }
Client is bound and listening { port: 54935 }
Server port: { port: 49934 }
Checking connection...
Got connection!
Live connected { type: 'start' }
```

When you load the plugin, you should also see the port that it uses to listen to messages from ableton-js in Live's status bar:

![CleanShot 2023-04-06 at 17 11 22](https://user-images.githubusercontent.com/541628/230422187-7debe843-5a97-45dd-b873-8d1a0b3795ba.png)
